lib/location/method_cloud_location.c

```diff
@@ -0,0 +1,441 @@
+		LOG_WRN("Wi-Fi scan request failed (%d).", status->status);
+	} else {
+		LOG_INF("Scan request done.");
+	}
+
+	latest_scan_result_count =
+		(current_scan_result_count > CONFIG_LOCATION_METHOD_WIFI_SCANNING_RESULTS_MAX_CNT) ?
+			CONFIG_LOCATION_METHOD_WIFI_SCANNING_RESULTS_MAX_CNT :
+			current_scan_result_count;
+	current_scan_result_count = 0;
+	k_sem_give(&wifi_scan_ready);
+}
+
+static struct net_mgmt_event_callback method_wifi_net_mgmt_cb;
+
+void method_cloud_location_net_mgmt_event_handler(
+	struct net_mgmt_event_callback *cb,
+	uint32_t mgmt_event,
+	struct net_if *iface)
+{
+	ARG_UNUSED(iface);
+
+	if (running) {
+		switch (mgmt_event) {
+		case NET_EVENT_WIFI_SCAN_RESULT:
+			method_wifi_scan_result_handle(cb);
+			break;
+		case NET_EVENT_WIFI_SCAN_DONE:
+			method_wifi_scan_done_handle(cb);
+			break;
+		default:
+			break;
+		}
+	}
+}
+
+/******************************************************************************/
+
+static void method_cloud_location_positioning_work_fn(struct k_work *work)
+{
+	struct location_cloud_serv_pos_req params = { 0 };
+	struct location_data location;
+	struct location_data location_result = { 0 };
+	struct method_cloud_location_start_work_args *work_data =
+		CONTAINER_OF(work, struct method_cloud_location_start_work_args, work_item);
+	const struct location_wifi_config *wifi_config = work_data->wifi_config;
+	const struct location_cellular_config *cell_config = work_data->cell_config;
+	int64_t scan_start_time;
+	int64_t scan_time;
+	int32_t used_timeout_ms;
+	int err = 0;
+
+	if (wifi_config && cell_config) {
+		used_timeout_ms = MIN(cell_config->timeout, wifi_config->timeout);
+	} else if (cell_config) {
+		used_timeout_ms = cell_config->timeout;
+	} else if (wifi_config) {
+		used_timeout_ms = wifi_config->timeout;
+	} else {
+		__ASSERT_NO_MSG(false);
+	}
+
+	location_core_timer_start(used_timeout_ms);
+	scan_start_time = k_uptime_get();
+
+	if (wifi_config) {
+		LOG_INF("Triggering WiFi Scanning");
+		err = method_wifi_scanning_start();
+		if (err) {
+			LOG_WRN("Cannot start Wi-Fi scanning, err %d", err);
+		}
+	} else {
+		k_sem_give(&wifi_scan_ready);
+	}
+
+	if (cell_config) {
+		LOG_INF("Triggering neighbor cell measurements");
+		err = method_cellular_ncellmeas_start();
+		if (err) {
+			LOG_WRN("Cannot start neighbor cell measurements");
+		}
+	} else {
+		k_sem_give(&ncellmeas_ready);
+	}
+
+	k_sem_take(&wifi_scan_ready, K_FOREVER);
+	k_sem_take(&ncellmeas_ready, K_FOREVER);
+	if (!running) {
+		goto end;
+	}
+
+	location_core_timer_stop();
+
+	/* Scannings done at this point of time. Store current time to response. */
+	location_utils_systime_to_location_datetime(&location_result.datetime);
+
+	/* Check if timeout is given */
+	params.timeout_ms = used_timeout_ms;
+	if (used_timeout_ms != SYS_FOREVER_MS) {
+		/* +1 to round the time up */
+		scan_time = (k_uptime_get() - scan_start_time) + 1;
+
+		/* Check if timeout has already elapsed */
+		if (scan_time >= used_timeout_ms) {
```

Comment:
Shouldn't the timeout be an argument to k_sem_take on lines 277 and 278? Or more precisely, the timeout of second k_sem_take would be whatever is left after k_sem_take  (if any)

Reply:
`location_core_timer_start(used_timeout_ms)` handles this on line 254 and we don't want duplicate timers as there would be race conditions between them.